File: 1016.cc
#include "stdio.h"
#include "string"
int main()
{
    int i=0, j=0, k=0;
    char strs[81]={0};
    while(scanf("%s", &strs) && strs[0]!='-')
    {
        char loop[16][10]={0}, tmp[31]={0}, tmpInt[3]={0};
        for(printf("%s ", strs), i=0; strs[i]!=0 && ++loop[0][strs[i]-48]; i++);
        for(i=0; i<10 && (loop[0][i]!=0 && (strcat(tmp, (sprintf(tmpInt, "%d", (int)loop[0][i]), tmpInt)), strcat(tmp, (sprintf(tmpInt, "%d", i), tmpInt))), true); i++);
        if(!strcmp(strs, tmp) && printf("is self-inventorying\n"))    continue;
        for(j=1; j<15; j++)
        {
            for(i=0; tmp[i]!=0 && ++loop[j][tmp[i]-48]; i++);
            for(i=0; i<j; i++)
            {
                for(k=0; k<10 && loop[i][k]==loop[j][k]; k++);
                if(k==10 && (j-i==1 ? printf("is self-inventorying after %d steps\n", j) : printf("enters an inventory loop of length %d\n", j-i)))
                    break;
            }
            if(i!=j)    break;
            for(i=0, tmp[0]=0; i<10 && (loop[j][i]!=0 && (strcat(tmp, (sprintf(tmpInt, "%d", (int)loop[j][i]), tmpInt)), strcat(tmp, (sprintf(tmpInt, "%d", i), tmpInt))), true); i++);
        }
        j==15 && printf("can not be classified after 15 iterations\n");
    }
    return 0;
}

File: 1016.java
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStreamReader;
 import java.util.Arrays;

 public class Main {

     public static void main(String[] args) throws IOException {
         BufferedReader read = new BufferedReader(new InputStreamReader(
                 System.in));
         String s;
         char[][] c;
         char[] next;
         int step;
         int[] num;
         int result;
         StringBuffer buff = new StringBuffer();
         while (!(s = read.readLine()).startsWith("-")) {
             c = new char[15][];
             step = 2;
             num = new int[10];
             c[0] = s.toCharArray();
             while (true) {
                 Arrays.fill(num, 0);
                 for (int i = 0; i < c[step - 2].length; i++) {
                     num[c[step - 2][i] - '0']++;
                 }
                 buff.delete(0, buff.length());
                 for (int i = 0; i < 10; i++) {
                     if (num[i] != 0) {
                         buff.append(num[i]);
                         buff.append(i);
                     }
                 }
                 next = buff.toString().toCharArray();
                 if ((result = equal(c, next, step)) != -1) {
                     if (step == 2) {
                         System.out.printf("%s is self-inventorying\n", s);
                     } else {
                         if (step - result == 2) {
                             System.out.printf(
                                     "%s is self-inventorying after %d steps\n",
                                     s, step - 1);
                         } else {
                             System.out
                                     .printf(
                                             "%s enters an inventory loop of length %d\n",
                                             s, step - result - 1);
                         }
                     }
                     break;
                 } else {
                     if (step >= 15) {
                         System.out
                                 .printf(
                                         "%s can not be classified after 15 iterations\n",
                                         s);
                         break;
                     }
                     c[step - 1] = next;
                     step++;
                 }
             }
         }
     }

     public static int equal(char[][] a, char[] b, int t) {
         int[] aa = new int[10];
         int[] bb = new int[10];
         loop: for (int i = 0; i < t - 1; i++) {
             Arrays.fill(aa, 0);
             Arrays.fill(bb, 0);
             if (a[i].length != b.length) {
                 continue;
             } else {
                 for (int j = 0; j < b.length; j++) {
                     aa[a[i][j] - '0']++;
                     bb[b[j] - '0']++;
                 }
                 for (int j = 0; j < 10; j++) {
                     if (aa[j] != bb[j]) {
                         continue loop;
                     }
                 }
                 return i;
             }
         }
         return -1;
     } 
}